plrust/src/plrust_proc.rs

```diff
@@ -105,19 +99,55 @@ pub(crate) fn load(pg_proc_oid: pg_sys::Oid) -> eyre::Result<UserCrate<StateLoad
     Ok(loaded)
 }
 
-/// helper function to build a vec of Spi arguments to be used as the composite primary key
-/// `plrust.plrust_proc` needs to locate a function
+// helper function to build the primary key values used to query `plrust.plrust_proc` via Spi
+#[rustfmt::skip]
 #[inline]
 fn pkey_datums(pg_proc_oid: pg_sys::Oid) -> Vec<(PgOid, Option<pg_sys::Datum>)> {
     vec![
-        (PgBuiltInOids::REGPROCOID.oid(), pg_proc_oid.into_datum()),
-        (
-            PgBuiltInOids::TEXTOID.oid(),
-            get_target_triple().into_datum(),
-        ),
+        get_fn_identity_datum(pg_proc_oid),
+        (PgBuiltInOids::TEXTOID.oid(), get_target_triple().into_datum()),
     ]
 }
 
+// helper function to build the function identity (oid, value) datum
+fn get_fn_identity_datum(pg_proc_oid: pg_sys::Oid) -> (PgOid, Option<pg_sys::Datum>) {
+    let oa = pg_sys::ObjectAddress {
+        classId: pg_sys::ProcedureRelationId,
+        objectId: pg_proc_oid,
+        objectSubId: 0,
+    };
+    let identity_ptr = unsafe {
+        #[cfg(feature = "pg13")]
+        {
+            // SAFETY:  getObjectIdentity will raise an ERROR if the ObjectAddress we created doesn't
+            // exist, otherwise it returns a properly palloc'd pointer
+            pg_sys::getObjectIdentity(&oa as *const _)
+        }
+
+        #[cfg(not(feature = "pg13"))]
+        {
+            // SAFETY:  by setting "missing_ok" to false, getObjectIdentity will raise an ERROR if the
+            // ObjectAddress we created doesn't exist, otherwise it returns a properly palloc'd pointer
+            pg_sys::getObjectIdentity(&oa as *const _, false)
+        }
+    };
+    let identity_str = unsafe {
+        // SAFETY:  Postgres has given us a valid, albeit palloc'd, cstring as the result of getObjectIdentity
+        CStr::from_ptr(identity_ptr).to_str().unwrap_or_else(|_| {
+            pgx::error!("function {pg_proc_oid}'s identity is not a valid UTF8 string")
+        })
+    };
+
+    let result = (PgBuiltInOids::TEXTOID.oid(), identity_str.into_datum());
+
+    unsafe {
+        // SAFETY: identity_ptr was previously proven valid
+        pg_sys::pfree(identity_ptr.cast());
+    }
```

Comment:
I think this has another hidden assumption
```suggestion
    unsafe {
        // SAFETY: identity_ptr was previously proven valid and
        // identity_str was reallocated elsewhere in Postgres
        pg_sys::pfree(identity_ptr.cast());
    }
```